README.md
Chronic
=======

## DESCRIPTION

Chronic is a natural language date/time parser written in pure Ruby. See below
for the wide variety of formats Chronic will parse.


## INSTALLATION

### RubyGems

    $ [sudo] gem install chronic

### GitHub

    $ git clone git://github.com/mojombo/chronic.git
    $ cd chronic && gem build chronic.gemspec
    $ gem install chronic-<version>.gem


## USAGE

You can parse strings containing a natural language date using the
`Chronic.parse` method.

    require 'chronic'

    Time.now   #=> Sun Aug 27 23:18:25 PDT 2006

    Chronic.parse('tomorrow')
      #=> Mon Aug 28 12:00:00 PDT 2006

    Chronic.parse('monday', :context => :past)
      #=> Mon Aug 21 12:00:00 PDT 2006

    Chronic.parse('this tuesday 5:00')
      #=> Tue Aug 29 17:00:00 PDT 2006

    Chronic.parse('this tuesday 5:00', :ambiguous_time_range => :none)
      #=> Tue Aug 29 05:00:00 PDT 2006

    Chronic.parse('may 27th', :now => Time.local(2000, 1, 1))
      #=> Sat May 27 12:00:00 PDT 2000

    Chronic.parse('may 27th', :guess => false)
      #=> Sun May 27 00:00:00 PDT 2007..Mon May 28 00:00:00 PDT 2007
      
    Chronic.parse('6/4/2012', :endian_precedence => :little)
      #=> Fri Apr 06 00:00:00 PDT 2012
      

See `Chronic.parse` for detailed usage instructions.


## EXAMPLES

Chronic can parse a huge variety of date and time formats. Following is a
small sample of strings that will be properly parsed. Parsing is case
insensitive and will handle common abbreviations and misspellings.

Simple

* thursday
* november
* summer
* friday 13:00
* mon 2:35
* 4pm
* 6 in the morning
* friday 1pm
* sat 7 in the evening
* yesterday
* today
* tomorrow
* this tuesday
* next month
* last winter
* this morning
* last night
* this second
* yesterday at 4:00
* last friday at 20:00
* last week tuesday
* tomorrow at 6:45pm
* afternoon yesterday
* thursday last week

Complex

* 3 years ago
* a year ago
* 5 months before now
* 7 hours ago
* 7 days from now
* 1 week hence
* in 3 hours
* 1 year ago tomorrow
* 3 months ago saturday at 5:00 pm
* 7 hours before tomorrow at noon
* 3rd wednesday in november
* 3rd month next year
* 3rd thursday this september
* 4th day last week
* fourteenth of june 2010 at eleven o'clock in the evening
* may seventh '97 at three in the morning

Specific Dates

* January 5
* 22nd of june
* 5th may 2017
* February twenty first
* dec 25
* may 27th
* October 2006
* oct 06
* jan 3 2010
* february 14, 2004
* february 14th, 2004
* 3 jan 2000
* 17 april 85
* 5/27/1979
* 27/5/1979
* 05/06
* 1979-05-27
* Friday
* 5
* 4:00
* 17:00
* 0800

Specific Times (many of the above with an added time)

* January 5 at 7pm
* 22nd of june at 8am
* 1979-05-27 05:00:00
* etc


## TIME ZONES

Chronic allows you to set which Time class to use when constructing times. By
default, the built in Ruby time class creates times in your system's local
time zone. You can set this to something like ActiveSupport's
[TimeZone](http://api.rubyonrails.org/classes/ActiveSupport/TimeZone.html)
class to get full time zone support.

    >> Time.zone = "UTC"
    >> Chronic.time_class = Time.zone
    >> Chronic.parse("June 15 2006 at 5:45 AM")
    => Thu, 15 Jun 2006 05:45:00 UTC +00:00


## Localization

Chronic supports basic localization. You can either create your own based on
the default english locale or use the
[chronic-l10n](http://github.com/luan/chronic-l10n) gem. You can change the
locale by simply changing the `Chronic.locale` value, such as (given you have
:pt-BR available):

    >> Chronic.locale = :'pt-BR'
    >> Chronic.parse("15 de Junho de 2006 as 5:45 da manha")
    => Thu, 15 Jun 2006 05:45:00 UTC +00:00


## LIMITATIONS

Chronic uses Ruby's built in Time class for all time storage and computation.
Because of this, only times that the Time class can handle will be properly
parsed. Parsing for times outside of this range will simply return nil.
Support for a wider range of times is planned for a future release.


## CONTRIBUTE

If you'd like to hack on Chronic, start by forking the repo on GitHub:

https://github.com/mojombo/chronic

The best way to get your changes merged back into core is as follows:

1. Clone down your fork
1. Create a thoughtfully named topic branch to contain your change
1. Hack away
1. Add tests and make sure everything still passes by running `rake`
1. Ensure your tests pass in multiple timezones. ie `TZ=utc rake` `TZ=BST rake`
1. If you are adding new functionality, document it in the README
1. Do not change the version number, we will do that on our end
1. If necessary, rebase your commits into logical chunks, without errors
1. Push the branch up to GitHub
1. Send a pull request for your branch


HISTORY.md
# HEAD

* Added localization support

# 0.7.0 / 2012-08-01

* Support parsing EXIF date format (#112)
* Start using minitest for testing
* Ensure periods are interpreted as colons (#81).
* Support month/day and day/month parsing (#59).
* Support day(scalar)-month(name)-year(scalar) (#99).
* Handle text starting with 'a' or 'an' (#101, @steveburkett).
* Ensure post medium timestamps are correctly formatted (#89)

# 0.6.7 / 2012-01-31

* Handle day, month names with scalar day and year (Joe Fiorini)
* Ensure 31st parses correctly with day names (Joe Fiorini)

# 0.6.6 / 2011-11-23

* `Chronic.parse('thur')` no longer returns `nil` (@harold)

# 0.6.5 / 2011-11-04

* Fix bug when parsing ordinal repeaters (#73)
* Added handler support for day_name month_name (@imme5150)
* Fix bug when parsing strings prefixed with PM

# 0.6.4 / 2011-08-08

* Fixed bug where 'noon' was parsed as 00:00 rather than 12:00
  with :ambiguous_time_range => :none (Vladimir Chernis)
* Add support for handling '2009 May 22nd'
* Add the ability to handle scalar-day/repeater-month-name as well as ordinals

# 0.6.3 / 2011-08-01

* Ensure 'thu' is parsed as Thursday for 1.8.7 generic timestamp

# 0.6.2 / 2011-07-28

* Ensure specific endian handlers are prioritised over normal date handlers
* Recognize UTC as timezone and accept HH::MM timezone offset (Jason Dusek)

# 0.6.1 / 2011-07-21

* Ensure Handler definitions are executed in the correct order

# 0.6.0 / 2011-07-19

* Attempting to parse strings with days past the last day of a month will
  now return nil. ex: `Chronic.parse("30th February") #=> nil`
* All deprecated methods are marked for removal in Chronic 0.7.0
* Deprecated `Chronic.numericize_numbers` instead use
  `Chronic::Numerizer.numerize`
* Deprecated `Chronic::InvalidArgumentException` and instead use
  `ArgumentError`
* Deprecated `Time.construct` and use `Chronic.construct` in place of this
* Deprecated `Time#to_minidate`, instead use `Chronic::MiniDate.from_time(time)`
* Add support for handling generic timestamp for Ruby 1.9+

# 0.5.0 / 2011-07-01

* Replace commas with spaces instead of removing the char (Thomas Walpole)
* Added tests for RepeaterSeason
* Re-factored tests. Now rather than having a test_parsing method for testing
  all handlers, break them down independent of handler method. For example
  with handler `handle_sm_sd_sy` the subsequent test would be
  `test_handle_sm_sd_sy`
* Added support for parsing ordinal-dates/month-names/year, ie:
  `2nd of May 1995`
* Added support for parsing ordinal-dates and month names, ie:
  `22nd of February at 6:30pm`
* Fix `Time.construct` leap year checking. Instead use `Date.leap?(year)`

# 0.4.4 / 2011-06-12

* Fix RepeaterYear for fetching past year offsets when the current day is
  later than the last day of the same month in a past year (leap years) ie
  on 29th/feb (leap year) `last year` should (and now does) return 28th/feb
  instead of 1st/march
* Opt in for gem testing http://test.rubygems.org/

# 0.4.3 / 2011-06-08

* Fix issue with parsing 1:xxPM -- Ensure 1 is treated as ambiguous, not
  just >1

# 0.4.2 / 2011-06-07

* Fix MonthRepeater for fetching past month offsets when current day is
  later than the last day of a past month (ie on 29th of March when parsing
  `last month` Chronic would return March instead of February. Now Chronic
  returns the last day of the past month)

# 0.4.1 / 2011-06-05

* Fix MiniDate ranges for parsing seasons (Thomas Walpole)

# 0.4.0 / 2011-06-04

* Ensure context is being passed through grabbers. Now "Sunday at 2:18pm"
  with `:context => :past` will return the correct date
* Support parsing ordinal strings (eg first, twenty third => 1st, 23rd)
* Seasons now ignore DST and return 00 as an hour
* Support parsing 2 digit years and added `ambiguous_year_future_bias` option
* Support parsing 'thurs' for Thursday
* Fix pre_normalize() to remove periods before numerizing
* Fix RepeaterDays to not add an extra hour in future tense. This meant
  when parsing 'yesterday' after 11PM, Chronic would return today
* Discard any prefixed 0 for time strings when using post noon portion
* Gemspec updates for RubyGems deprecations
* Ensure 0:10 is treated like 00:10
* Ensure we load classes after setting Chronic class instance variables
  so we can debug initialization and do assignments at compile time
* Added a Tag.scan_for method for DRYing up some scanning code
* Move some classes into their own files for maintainability
* Numerizer.andition should be a private class method, make it so
* Namespaced Numerizer, Season and MiniDate (Sascha Teske)
* Support for Ruby 1.9 (Dave Lee, Aaron Hurley)
* Fix `:context => :past` where parsed date is in current month (Marc Hedlund)
* Fix undefined variable in RepeaterHour (Ryan Garver)
* Added support for parsing 'Fourty' as another mis-spelling (Lee Reilly)
* Added ordinal format support: ie 'February 14th, 2004' (Jeff Felchner)
* Fix dates when working with daylight saving times (Mike Mangino)

# 0.3.0 / 2010-10-22

* Fix numerizer number combination bug (27 Oct 2006 7:30pm works now)
* Allow numeric timezone offset (e.g -0500)
* Disregard commas (so as to not return nil)
* Fix parse of (am|pm|oclock) separation to handle "Ham sandwich" properly
* Handle 'on' e.g. 5pm on Monday
* Support seasons
* Support weekend/weekday
* Add endianness option
* Update version number in the module
* Fix/improve logic checks in Ordinal, and Scalar
* Parse 'a' or 'p' as 'am' and 'pm' google-calendar style
* Dates < 1 are not valid
* Fix bugs related to timezone offset
* Use RakeGem for build management
* Reformat README and HISTORY to use Markdown
* Global whitespace removal

# 0.2.3

* Fix 12am/12pm

# 0.2.2

* Add missing files (damn you manifest)

# 0.2.1

* Fix time overflow issue
* Implement "next" for minute repeater
* Generalize time dealiasing to dealias regardless of day portion and
  time position
* Add additional token match for cases like "friday evening at 7" and

lib/chronic/chronic.rb
module Chronic

  # Returns a Hash of default configuration options.
  DEFAULT_OPTIONS = {
    :context => :future,
    :now => nil,
    :guess => true,
    :ambiguous_time_range => 6,
    :endian_precedence    => [:middle, :little],
    :ambiguous_year_future_bias => 50
  }

  class << self

    # Parses a string containing a natural language date or time.
    #
    # If the parser can find a date or time, either a Time or Chronic::Span
    # will be returned (depending on the value of `:guess`). If no
    # date or time can be found, `nil` will be returned.
    #
    # text - The String text to parse.
    # opts - An optional Hash of configuration options:
    #        :context - If your string represents a birthday, you can set
    #                   this value to :past and if an ambiguous string is
    #                   given, it will assume it is in the past.
    #        :now - Time, all computations will be based off of time
    #               instead of Time.now.
    #        :guess - By default the parser will guess a single point in time
    #                 for the given date or time. If you'd rather have the
    #                 entire time span returned, set this to false
    #                 and a Chronic::Span will be returned.
    #        :ambiguous_time_range - If an Integer is given, ambiguous times
    #                  (like 5:00) will be assumed to be within the range of
    #                  that time in the AM to that time in the PM. For
    #                  example, if you set it to `7`, then the parser will
    #                  look for the time between 7am and 7pm. In the case of
    #                  5:00, it would assume that means 5:00pm. If `:none`
    #                  is given, no assumption will be made, and the first
    #                  matching instance of that time will be used.
    #        :endian_precedence - By default, Chronic will parse "03/04/2011"
    #                 as the fourth day of the third month. Alternatively you
    #                 can tell Chronic to parse this as the third day of the
    #                 fourth month by setting this to [:little, :middle].
    #        :ambiguous_year_future_bias - When parsing two digit years
    #                 (ie 79) unlike Rubys Time class, Chronic will attempt
    #                 to assume the full year using this figure. Chronic will
    #                 look x amount of years into the future and past. If the
    #                 two digit year is `now + x years` it's assumed to be the
    #                 future, `now - x years` is assumed to be the past.
    #
    # Returns a new Time object, or Chronic::Span if :guess option is false.
    def parse(text, opts={})
      # ensure current locale is available
      raise ArgumentError, "#{locale} is not an available locale" unless has_locale(locale)
      options = DEFAULT_OPTIONS.merge opts

      # ensure the specified options are valid
      (opts.keys - DEFAULT_OPTIONS.keys).each do |key|
        raise ArgumentError, "#{key} is not a valid option key."
      end

      unless [:past, :future, :none].include?(options[:context])
        raise ArgumentError, "Invalid context, :past/:future only"
      end

      options[:text] = text
      Chronic.now = options[:now] || Chronic.time_class.now

      # tokenize words
      tokens = tokenize(text, options)

      if Chronic.debug
        puts "+#{'-' * 51}\n| #{tokens}\n+#{'-' * 51}"
      end

      span = tokens_to_span(tokens, options)

      if span
        options[:guess] ? guess(span) : span
      end
    end

    # Clean up the specified text ready for parsing.
    #
    # Clean up the string by stripping unwanted characters, converting
    # idioms to their canonical form, converting number words to numbers
    # (three => 3), and converting ordinal words to numeric
    # ordinals (third => 3rd)
    #
    # text - The String text to normalize.
    #
    # Examples:
    #
    #   Chronic.pre_normalize('first day in May')
    #     #=> "1st day in may"
    #
    #   Chronic.pre_normalize('tomorrow after noon')
    #     #=> "next day future 12:00"
    #
    #   Chronic.pre_normalize('one hundred and thirty six days from now')
    #     #=> "136 days future this second"
    #
    # Returns a new String ready for Chronic to parse.
    def pre_normalize(text)
      text = text.to_s.downcase
      text = translate([:pre_normalize, :preprocess]).call(text)
      translate([:pre_normalize, :pre_numerize]).each do |sub|
        text.gsub!(*sub)
      end
      text = Numerizer.numerize(text)
      translate([:pre_normalize, :pos_numerize]).each do |sub|
        text.gsub!(*sub)
      end
      text
    end

    # Convert number words to numbers (three => 3, fourth => 4th).
    #
    # text - The String to convert.
    #
    # Returns a new String with words converted to numbers.
    def numericize_numbers(text)
      warn "Chronic.numericize_numbers will be deprecated in version 0.7.0. Please use Chronic::Numerizer.numerize instead"
      Numerizer.numerize(text)
    end

    # Guess a specific time within the given span.
    #
    # span - The Chronic::Span object to calcuate a guess from.
    #
    # Returns a new Time object.
    def guess(span)
      if span.width > 1
        span.begin + (span.width / 2)
      else
        span.begin
      end
    end

    # List of Handler definitions. See #parse for a list of options this
    # method accepts.
    #
    # options - An optional Hash of configuration options:
    #           :endian_precedence -
    #
    # Returns A Hash of Handler definitions.
    def definitions(options={})
      options[:endian_precedence] ||= [:middle, :little]

      @definitions ||= {
        :time => [
          Handler.new([:repeater_time, :repeater_day_portion?], nil)
        ],

        :date => [
          Handler.new([:repeater_day_name, :repeater_month_name, :scalar_day, :repeater_time, :separator_slash_or_dash?, :time_zone, :scalar_year], :handle_rdn_rmn_sd_t_tz_sy),
          Handler.new([:repeater_day_name, :repeater_month_name, :scalar_day], :handle_rdn_rmn_sd),
          Handler.new([:repeater_day_name, :repeater_month_name, :scalar_day, :scalar_year], :handle_rdn_rmn_sd_sy),
          Handler.new([:repeater_day_name, :repeater_month_name, :ordinal_day], :handle_rdn_rmn_od),
          Handler.new([:scalar_year, :separator_slash_or_dash, :scalar_month, :separator_slash_or_dash, :scalar_day, :repeater_time, :time_zone], :handle_sy_sm_sd_t_tz),
          Handler.new([:repeater_month_name, :scalar_day, :scalar_year], :handle_rmn_sd_sy),
          Handler.new([:repeater_month_name, :ordinal_day, :scalar_year], :handle_rmn_od_sy),
          Handler.new([:repeater_month_name, :scalar_day, :scalar_year, :separator_at?, 'time?'], :handle_rmn_sd_sy),
          Handler.new([:repeater_month_name, :ordinal_day, :scalar_year, :separator_at?, 'time?'], :handle_rmn_od_sy),
          Handler.new([:repeater_month_name, :scalar_day, :separator_at?, 'time?'], :handle_rmn_sd),
          Handler.new([:repeater_time, :repeater_day_portion?, :separator_on?, :repeater_month_name, :scalar_day], :handle_rmn_sd_on),
          Handler.new([:repeater_month_name, :ordinal_day, :separator_at?, 'time?'], :handle_rmn_od),
          Handler.new([:ordinal_day, :repeater_month_name, :scalar_year, :separator_at?, 'time?'], :handle_od_rmn_sy),
          Handler.new([:ordinal_day, :repeater_month_name, :separator_at?, 'time?'], :handle_od_rmn),
          Handler.new([:scalar_year, :repeater_month_name, :ordinal_day], :handle_sy_rmn_od),
          Handler.new([:repeater_time, :repeater_day_portion?, :separator_on?, :repeater_month_name, :ordinal_day], :handle_rmn_od_on),
          Handler.new([:repeater_month_name, :scalar_year], :handle_rmn_sy),
          Handler.new([:scalar_day, :repeater_month_name, :scalar_year, :separator_at?, 'time?'], :handle_sd_rmn_sy),
          Handler.new([:scalar_day, :repeater_month_name, :separator_at?, 'time?'], :handle_sd_rmn),
          Handler.new([:scalar_year, :separator_slash_or_dash, :scalar_month, :separator_slash_or_dash, :scalar_day, :separator_at?, 'time?'], :handle_sy_sm_sd),
          Handler.new([:scalar_month, :separator_slash_or_dash, :scalar_day], :handle_sm_sd),
          Handler.new([:scalar_month, :separator_slash_or_dash, :scalar_year], :handle_sm_sy),
          Handler.new([:scalar_day, :separator_slash_or_dash, :repeater_month_name, :separator_slash_or_dash, :scalar_year], :handle_sm_rmn_sy)
        ],

        # tonight at 7pm
        :anchor => [
          Handler.new([:grabber?, :repeater, :separator_at?, :repeater?, :repeater?], :handle_r),
          Handler.new([:grabber?, :repeater, :repeater, :separator_at?, :repeater?, :repeater?], :handle_r),
          Handler.new([:repeater, :grabber, :repeater], :handle_r_g_r)
        ],

        # 3 weeks from now, in 2 months
        :arrow => [
          Handler.new([:scalar, :repeater, :pointer], :handle_s_r_p),
          Handler.new([:pointer, :scalar, :repeater], :handle_p_s_r),
          Handler.new([:scalar, :repeater, :pointer, 'anchor'], :handle_s_r_p_a)
        ],

        # 3rd week in march
        :narrow => [
          Handler.new([:ordinal, :repeater, :separator_in, :repeater], :handle_o_r_s_r),
          Handler.new([:ordinal, :repeater, :grabber, :repeater], :handle_o_r_g_r)
        ]
      }

      endians = [
        Handler.new([:scalar_month, :separator_slash_or_dash, :scalar_day, :separator_slash_or_dash, :scalar_year, :separator_at?, 'time?'], :handle_sm_sd_sy),
        Handler.new([:scalar_day, :separator_slash_or_dash, :scalar_month, :separator_slash_or_dash, :scalar_year, :separator_at?, 'time?'], :handle_sd_sm_sy)
      ]

      case endian = Array(options[:endian_precedence]).first
      when :little
        @definitions[:endian] = endians.reverse
      when :middle
        @definitions[:endian] = endians
      else
        raise ArgumentError, "Unknown endian option '#{endian}'"
      end

      @definitions
    end

    # Construct a new time object determining possible month overflows
    # and leap years.
    #
    # year   - Integer year.
    # month  - Integer month.
    # day    - Integer day.
    # hour   - Integer hour.
    # minute - Integer minute.
    # second - Integer second.
    #
    # Returns a new Time object constructed from these params.
    def construct(year, month = 1, day = 1, hour = 0, minute = 0, second = 0)
      if second >= 60
        minute += second / 60
        second = second % 60
      end

      if minute >= 60
        hour += minute / 60
        minute = minute % 60
      end

      if hour >= 24
        day += hour / 24
        hour = hour % 24
      end

      # determine if there is a day overflow. this is complicated by our crappy calendar
      # system (non-constant number of days per month)
      day <= 56 || raise("day must be no more than 56 (makes month resolution easier)")
      if day > 28
        # no month ever has fewer than 28 days, so only do this if necessary
        leap_year_month_days = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        common_year_month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        days_this_month = Date.leap?(year) ? leap_year_month_days[month - 1] : common_year_month_days[month - 1]
        if day > days_this_month
          month += day / days_this_month
          day = day % days_this_month
        end
      end

      if month > 12
        if month % 12 == 0
          year += (month - 12) / 12
          month = 12
        else
          year += month / 12
          month = month % 12
        end
      end

      Chronic.time_class.local(year, month, day, hour, minute, second)
    end

    # The current locale Chronic is using to parse strings
    #
    # Examples:
    #
    #   require 'chronic'
    #
    #   Chronic.locale = :'pt-BR'
    #   Chronic.parse('15 de Junho de 2006 as 5:54 da manha ')
    #     # => Thu, 15 Jun 2006 05:45:00 UTC +00:00
    #
    # Returns the locale name Chronic uses internally
    def locale
      Thread.current[:locale] ||= :en
    end

    # Sets the current locale for the current thread
    #
    # value - Symbol locale
    def locale=(value)
      Thread.current[:locale] = value
    end

    # Adds a locale to the locale hash
    #
    # name - Symbol locale name
    # locale - Hash locale values
    def add_locale(name, locale)
      raise ArgumentError, "Locale shoud be a hash" unless locale.is_a?(Hash)
      locale_hashes[name] = locale
    end

    # Checks if a locale is available
    #
    # name - Symbol locale name
    #
    # Returns true if the locale is available, false if not
    def has_locale(name)
      locale_hashes.include? name
    end

    # Returns the translations for the current locale
    def translate(keys, loc=nil)
      loc ||= locale
      node = locale_hashes[loc]

      keys.each do |key|
        if node.include? key
          node = node[key]
        else
          return translate(keys, :en)
        end
      end

      node
    end

    private

    def tokenize(text, options)
      text = pre_normalize(text)
      tokens = text.split(' ').map { |word| Token.new(word) }
      [Repeater, Grabber, Pointer, Scalar, Ordinal, Separator, TimeZone].each do |tok|
        tok.scan(tokens, options)
      end
      tokens.select { |token| token.tagged? }
    end

    def tokens_to_span(tokens, options)
      definitions = definitions(options)

      (definitions[:endian] + definitions[:date]).each do |handler|
        if handler.match(tokens, definitions)
          good_tokens = tokens.select { |o| !o.get_tag Separator }
          return handler.invoke(:date, good_tokens, options)
        end
      end

      definitions[:anchor].each do |handler|
        if handler.match(tokens, definitions)
          good_tokens = tokens.select { |o| !o.get_tag Separator }
          return handler.invoke(:anchor, good_tokens, options)
        end
      end

      definitions[:arrow].each do |handler|
        if handler.match(tokens, definitions)
          good_tokens = tokens.reject { |o| o.get_tag(SeparatorAt) || o.get_tag(SeparatorSlashOrDash) || o.get_tag(SeparatorComma) }
          return handler.invoke(:arrow, good_tokens, options)
        end
      end

      definitions[:narrow].each do |handler|
        if handler.match(tokens, definitions)
          return handler.invoke(:narrow, tokens, options)
        end
      end

      puts "-none" if Chronic.debug
      return nil
    end

  end

  # Internal exception
  class ChronicPain < Exception
  end
end


lib/chronic/repeater.rb
module Chronic
  class Repeater < Tag

    # Scan an Array of Token objects and apply any necessary Repeater
    # tags to each token.
    #
    # tokens - An Array of tokens to scan.
    # options - The Hash of options specified in Chronic::parse.
    #
    # Returns an Array of tokens.
    def self.scan(tokens, options)
      tokens.each do |token|
        if t = scan_for_season_names(token) then token.tag(t); next end
        if t = scan_for_month_names(token) then token.tag(t); next end
        if t = scan_for_day_names(token) then token.tag(t); next end
        if t = scan_for_day_portions(token) then token.tag(t); next end
        if t = scan_for_times(token) then token.tag(t); next end
        if t = scan_for_units(token) then token.tag(t); next end
      end
    end

    # token - The Token object we want to scan.
    #
    # Returns a new Repeater object.
    def self.scan_for_season_names(token)
      scan_for token, RepeaterSeasonName, Chronic.translate([:repeater, :season_names])
    end

    # token - The Token object we want to scan.
    #
    # Returns a new Repeater object.
    def self.scan_for_month_names(token)
      scan_for token, RepeaterMonthName, Chronic.translate([:repeater, :month_names])
    end

    # token - The Token object we want to scan.
    #
    # Returns a new Repeater object.
    def self.scan_for_day_names(token)
      scan_for token, RepeaterDayName, Chronic.translate([:repeater, :day_names])
    end

    # token - The Token object we want to scan.
    #
    # Returns a new Repeater object.
    def self.scan_for_day_portions(token)
      scan_for token, RepeaterDayPortion, Chronic.translate([:repeater, :day_portions])
    end

    # token - The Token object we want to scan.
    #
    # Returns a new Repeater object.
    def self.scan_for_times(token)
      scan_for token, RepeaterTime, /^\d{1,2}(:?\d{2})?([\.:]?\d{2})?$/
    end

    # token - The Token object we want to scan.
    #
    # Returns a new Repeater object.
    def self.scan_for_units(token)
      Chronic.translate([:repeater, :units]).each do |item, symbol|
        if item =~ token.word
          klass_name = 'Repeater' + symbol.to_s.capitalize
          klass = Chronic.const_get(klass_name)
          return klass.new(symbol)
        end
      end
      return nil
    end

    def <=>(other)
      width <=> other.width
    end

    # returns the width (in seconds or months) of this repeatable.
    def width
      raise("Repeater#width must be overridden in subclasses")
    end

    # returns the next occurance of this repeatable.
    def next(pointer)
      raise("Start point must be set before calling #next") unless @now
    end

    def this(pointer)
      raise("Start point must be set before calling #this") unless @now
    end

    def to_s
      'repeater'
    end
  end
end


lib/chronic/locales/en.rb
module Chronic
  module Locales
    EN = {
      :pointer => {
        /\bpast\b/ => :past,
        /\b(?:future|in)\b/ => :future,
      },
      :ordinal_regex => /^(\d*)(st|nd|rd|th)$/,
      :numerizer => {
        :and => 'and',
        :preprocess => [
          [/ +|([^\d])-([^\d])/, '\1 \2'], # will mutilate hyphenated-words but shouldn't matter for date extraction
          [/a half/, 'haAlf'] # take the 'a' out so it doesn't turn into a 1, save the half for the end
        ],
        :fractional => [
          [/(\d+)(?: | and |-)*haAlf/i, proc { ($1.to_f + 0.5).to_s }]
        ],
        :direct_nums => [
          ['eleven', '11'],
          ['twelve', '12'],
          ['thirteen', '13'],
          ['fourteen', '14'],
          ['fifteen', '15'],
          ['sixteen', '16'],
          ['seventeen', '17'],
          ['eighteen', '18'],
          ['nineteen', '19'],
          ['ninteen', '19'], # Common mis-spelling
          ['zero', '0'],
          ['one', '1'],
          ['two', '2'],
          ['three', '3'],
          ['four(\W|$)', '4\1'],  # The weird regex is so that it matches four but not fourty
          ['five', '5'],
          ['six(\W|$)', '6\1'],
          ['seven(\W|$)', '7\1'],
          ['eight(\W|$)', '8\1'],
          ['nine(\W|$)', '9\1'],
          ['ten', '10'],
          ['\ba[\b^$]', '1'] # doesn't make sense for an 'a' at the end to be a 1
        ],
        :ordinals => [
          ['first', '1'],
          ['third', '3'],
          ['fourth', '4'],
          ['fifth', '5'],
          ['sixth', '6'],
          ['seventh', '7'],
          ['eighth', '8'],
          ['ninth', '9'],
          ['tenth', '10']
        ],
        :ten_prefixes => [
          ['twenty', 20],
          ['thirty', 30],
          ['forty', 40],
          ['fourty', 40], # Common mis-spelling
          ['fifty', 50],
          ['sixty', 60],
          ['seventy', 70],
          ['eighty', 80],
          ['ninety', 90]
        ],
        :big_prefixes => [
          ['hundred', 100],
          ['thousand', 1000],
          ['million', 1_000_000],
          ['billion', 1_000_000_000],
          ['trillion', 1_000_000_000_000],
        ],
      },

      :repeater => {
        :season_names => {
          /^springs?$/ => :spring,
          /^summers?$/ => :summer,
          /^(autumn)|(fall)s?$/ => :autumn,
          /^winters?$/ => :winter
        },
        :month_names => {
          /^jan\.?(uary)?$/ => :january,
          /^feb\.?(ruary)?$/ => :february,
          /^mar\.?(ch)?$/ => :march,
          /^apr\.?(il)?$/ => :april,
          /^may$/ => :may,
          /^jun\.?e?$/ => :june,
          /^jul\.?y?$/ => :july,
          /^aug\.?(ust)?$/ => :august,
          /^sep\.?(t\.?|tember)?$/ => :september,
          /^oct\.?(ober)?$/ => :october,
          /^nov\.?(ember)?$/ => :november,
          /^dec\.?(ember)?$/ => :december
        },
        :day_names => {
          /^m[ou]n(day)?$/ => :monday,
          /^t(ue|eu|oo|u|)s?(day)?$/ => :tuesday,
          /^we(d|dnes|nds|nns)(day)?$/ => :wednesday,
          /^th(u|ur|urs|ers)(day)?$/ => :thursday,
          /^fr[iy](day)?$/ => :friday,
          /^sat(t?[ue]rday)?$/ => :saturday,
          /^su[nm](day)?$/ => :sunday
        },
        :day_portions => {
          /^ams?$/ => :am,
          /^pms?$/ => :pm,
          /^mornings?$/ => :morning,
          /^afternoons?$/ => :afternoon,
          /^evenings?$/ => :evening,
          /^(night|nite)s?$/ => :night
        },
        :units => {
          /^years?$/ => :year,
          /^seasons?$/ => :season,
          /^months?$/ => :month,
          /^fortnights?$/ => :fortnight,
          /^weeks?$/ => :week,
          /^weekends?$/ => :weekend,
          /^(week|business)days?$/ => :weekday,
          /^days?$/ => :day,
          /^hrs?$/ => :hour,
          /^hours?$/ => :hour,
          /^mins?$/ => :minute,
          /^minutes?$/ => :minute,
          /^secs?$/ => :second,
          /^seconds?$/ => :second
        }
      },

      :pre_normalize => {
        :preprocess => proc {|str| str},
        :pre_numerize => [
          [/\./, ':'],
          [/['"]/, ''],
          [/,/, ' '],
          [/^second /, '2nd '],
          [/\bsecond (of|day|month|hour|minute|second)\b/, '2nd \1']
        ],
        :pos_numerize => [
          [/ \-(\d{4})\b/, ' tzminus\1'],
          [/([\/\-\,\@])/, ' \1 '],
          [/(?:^|\s)0(\d+:\d+\s*pm?\b)/, ' \1'],
          [/\btoday\b/, 'this day'],
          [/\btomm?orr?ow\b/, 'next day'],
          [/\byesterday\b/, 'last day'],
          [/\bnoon\b/, '12:00pm'],
          [/\bmidnight\b/, '24:00'],
          [/\bnow\b/, 'this second'],
          [/\b(?:ago|before(?: now)?)\b/, 'past'],
          [/\bthis (?:last|past)\b/, 'last'],
          [/\b(?:in|during) the (morning)\b/, '\1'],
          [/\b(?:in the|during the|at) (afternoon|evening|night)\b/, '\1'],
          [/\btonight\b/, 'this night'],
          [/\b\d+:?\d*[ap]\b/,'\0m'],
          [/(\d)([ap]m|oclock)\b/, '\1 \2'],
          [/\b(hence|after|from)\b/, 'future'],
          [/^\s?an? /i, '1 '],
          [/\b(\d{4}):(\d{2}):(\d{2})\b/, '\1 / \2 / \3'] # DTOriginal
        ]
      },

      :grabber => {
        /last/ => :last,
        /this/ => :this,
        /next/ => :next
      },

      :token => {
        :comma => /^,$/,
        :at => /^(at|@)$/,
        :in => /^in$/,
        :on => /^on$/
      }
    }
  end
end
